### ISSUE TITLE
chameneos.mlish fails tests

### ISSUE THREAD

**Issue by iitalics** [stchang/macrotypes#20]
chameneos is randomly failing tests. I don't know exactly what chameneos does but maybe we should do something about it?

**Comment by stchang** [comment]
Yes, it's an intermittent failure due to random inputs. I thought I had made the test deterministic but I must have missed a case. Feel free to fix it if can figure out a way.

chameneos is a standard benchmark in the language shootout benchmarks: http://benchmarksgame.alioth.debian.org/

I believe the original paper is here: http://ieeexplore.ieee.org/document/1227495/

**Comment by SuzanneSoy** [comment]
That `(sleep 0)` on [line 79](https://github.com/stchang/macrotypes/blob/master/turnstile/examples/tests/mlish/chameneos.mlish#L79) looks highly suspicious. The problem, though, is that the tests seem to more or less check whether the scheduler is (precisely) fair.

PR #30 fixes this by not relying on the scheduler.

**TL;DR:** *Although we could modify the benchmark to ensure that the creatures are scheduled in the right order, I think it's better to simply test against the sum of all meetings (knowing that there were 100 couple-meetings, there should be 200 meetings recorded by individual creatures), instead of checking that every creature got 66 or 67 meetings (which happens when they are run in the order `A B C A B C …`).*

Explanation of what I think was happening follows below.

-----



The `place` thread, which runs `swap`, gets two messages (with a callback channel on each) from the `ch-meet` channel, and replies to the callback by swapping the messages. The `creature`s send messages to `ch-meet`, and count how many other creatures they have met, and how many times they have met a creature with the same name (which could happen because the names are random).

The tests below check that each creature met an other one exactly 66 or 67 times, for 100 runs. This will occur if the messages are sent to `ch-meet` by the creatures `A B C A B C A B C …` in that order.

However, between the `(channel-put ch1 …) (channel-put ch2 …)` on lines [57–58](https://github.com/stchang/macrotypes/blob/master/turnstile/examples/tests/mlish/chameneos.mlish#L57) and the subsequent iteration which runs `(channel-get ch-meet) (channel-get ch-meet)` on lines [53](https://github.com/stchang/macrotypes/blob/master/turnstile/examples/tests/mlish/chameneos.mlish#L53) and [55](https://github.com/stchang/macrotypes/blob/master/turnstile/examples/tests/mlish/chameneos.mlish#L53), in principle any creature could grab the channel and send a message first (if the replies were just sent to `A` and `B`, I suspect it's possible that `C` was not being scheduled, and that `A` and `B` grab the channel again; it is also possible if `C` had already grabbed the channel in the meantime that `B` will come first, instead of `A`).

The results on [the benchmarks game page](http://benchmarksgame.alioth.debian.org/u64q/program.php?test=chameneosredux&lang=racket&id=1) for the Racket implementation show that the scheduling is not 100% fair (the creatures have 3999997, 4000011 and 3999992 meetings).

**Comment by stchang** [comment]
Fixed by #30
